pmd-test/src/main/java/net/sourceforge/pmd/testframework/RuleTst.java

```diff
@@ -555,4 +593,34 @@ private static String parseTextNode(Node exampleNode) {
         }
         return buffer.toString().trim();
     }
+
+    @TestFactory
+    Collection<DynamicTest> ruleTests() {
+        setUp();
+        final List<Rule> rules = new ArrayList<>(getRules());
+        rules.sort(Comparator.comparing(Rule::getName));
+
+        final List<TestDescriptor> tests = new LinkedList<>();
+        for (final Rule r : rules) {
+            final TestDescriptor[] ruleTests = extractTestsFromXml(r);
+            Collections.addAll(tests, ruleTests);
+        }
+
+        return tests.stream().map(this::toDynamicTest).collect(Collectors.toList());
+    }
+
+    private DynamicTest toDynamicTest(TestDescriptor testDescriptor) {
+        if (isIgnored(testDescriptor)) {
+            return DynamicTest.dynamicTest("[IGNORED] " + testDescriptor.getTestMethodName(),
+                    testDescriptor.getTestSourceUri(),
+                    () -> {});
+        }
+        return DynamicTest.dynamicTest(testDescriptor.getTestMethodName(),
+                testDescriptor.getTestSourceUri(),
```

Comment:
By providing a [testSourceUri](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests-uri-test-source) we can now double click in IntelliJ on the test and the xml test file opens and the correct line.
This unfortunately doesn't work in Eclipse :wink: 

Also we can right-click and run individual tests. Although for me, IntelliJ started the single rule test, but additionally all other tests in the module - something to investigate later.

Reply:
Oh, and I construct the testSourceUri obviously wrong under Windows

```
java.lang.IllegalArgumentException: Illegal character in opaque part at index 7: file:D:\a\pmd\pmd\pmd-apex\./src/test/resources/net/sourceforge/pmd/lang/apex/rule/security/xml/ApexXSSFromURLParam.xml?line=7
```

Reply:
This is great! To format the path correctly on windows I think the `Path` API would help

Reply:
This is fixed with https://github.com/pmd/pmd/pull/3945/commits/d7a673dec0e6369323e022d354a8c8b54c578d04